fuse_publishers/src/graph_publisher.cpp

```diff
@@ -0,0 +1,287 @@
+        if (variable_iter == index_by_variable_uuid_.end())
+        {
+          ROS_WARN_STREAM("Could not find variable " << fuse_core::uuid::to_string(variable.uuid())
+                                                     << " in the graph message.");
+          continue;
+        }
+
+        const auto& variable_index = variable_iter->second;
+        if (variable_index >= graph_msg_.variables.size())
+        {
+          ROS_WARN_STREAM("Variable " << fuse_core::uuid::to_string(variable.uuid()) << " index (" << variable_index
+                                      << ") is out of range. Graph message only has " << graph_msg_.variables.size()
+                                      << " variables.");
+          continue;
+        }
+
+        auto& variable_data = graph_msg_.variables[variable_index].data;
+        if (variable_data.size() != variable.size())
+        {
+          ROS_WARN_STREAM("Variable " << fuse_core::uuid::to_string(variable.uuid()) << " data size ("
+                                      << variable.size() << ") is different that the graph message variable size ("
+                                      << variable_data.size() << ").");
+          continue;
+        }
+
+        std::copy_n(variable.data(), variable.size(), variable_data.begin());
+      }
+    }
+
+    graph_publisher_.publish(graph_msg_);
+  }
+}
+
+void GraphPublisher::updateGraphMessage(fuse_core::Transaction::ConstSharedPtr transaction)
+{
+  for (const auto& variable : transaction->addedVariables())
+  {
+    addVariable(variable);
+  }
+
+  for (const auto& constraint : transaction->addedConstraints())
+  {
+    addConstraint(constraint);
+  }
+
+  for (const auto& constraint_uuid : transaction->removedConstraints())
+  {
+    removeConstraint(constraint_uuid);
+  }
+
+  for (const auto& variable_uuid : transaction->removedVariables())
+  {
+    removeVariable(variable_uuid);
+  }
+}
+
+fuse_msgs::Variable toMsg(const fuse_core::Variable& variable)
+{
+  fuse_msgs::Variable variable_msg;
+
+  variable_msg.uuid = fuse_core::uuid::to_string(variable.uuid());
+  variable_msg.type = variable.type();
+  std::copy_n(variable.data(), variable.size(), std::back_inserter(variable_msg.data));
+
+  return variable_msg;
+}
+
+fuse_msgs::Constraint toMsg(const fuse_core::Constraint& constraint)
+{
+  fuse_msgs::Constraint constraint_msg;
+
+  constraint_msg.uuid = fuse_core::uuid::to_string(constraint.uuid());
+  constraint_msg.type = constraint.type();
+
+  const auto& constraint_variables = constraint.variables();
+  constraint_msg.variables.reserve(constraint_variables.size());
+  std::transform(constraint_variables.begin(), constraint_variables.end(), std::back_inserter(constraint_msg.variables),
+                 [](const auto& variable_uuid) -> std::string { return fuse_core::uuid::to_string(variable_uuid); });
+
+  // FIXME populate the properties vector somehow from the cost function or maybe extend the API to support this
```

Comment:
I still don't have a good answer for this. Actually for the constraints "data", that I don't know how to serialize because it's embedded in the virtual cost function and only some child classes has methods to access it, like `mean` and `covariance`.